Adds command `spo web applicationcustomizer add`

MathijsVerbeeck opened the pull request:
Closes #3465

milanholemans commented:
Great work @MathijsVerbeeck!
We'll review it ASAP!

martinlingstuyl commented:
Awesome @MathijsVerbeeck, once the specs are clear, the work is easy :)

waldekmastykarz commented on docs/docs/cmd/spo/web/web-applicationcustomizer-add.md:
```diff
@@ -0,0 +1,56 @@
+
+## Remarks
+
+Running this command from the Windows Command Shell (cmd.exe) or PowerShell for Windows OS XP, 7, 8, 8.1 without bash installed might require additional formatting for command options that have JSON, XML or JavaScript values because the command shell treat quotes differently. For example, this is how ApplicationCustomizer user custom action can be created from the Windows cmd.exe:
```
We shouldn't mention the custom action here, because we're talking specifically about application customizer

waldekmastykarz replied:
In the command's description, we wrote `application customizer` while here we're writing `ApplicationCustomizer`. Let's use the former throughout.

waldekmastykarz commented on docs/docs/cmd/spo/web/web-applicationcustomizer-add.md:
````diff
@@ -0,0 +1,56 @@
+Adds an application customizer to the sales site.
+
+```sh
+m365 spo web applicationcustomizer add --title 'Some customizer' --clientSideComponentId  799883f5-7962-4384-a10a-105adaec6ffc --webUrl https://contoso.sharepoint.com/sites/sales
````
Let's remove double space in the value of the `clientSideComponentId` option

waldekmastykarz commented on docs/docs/cmd/spo/web/web-applicationcustomizer-add.md:
````diff
@@ -0,0 +1,56 @@
+Adds an application customizer to the sales site with some properties.
+
+```sh
+m365 spo web applicationcustomizer add --title 'Some customizer' --clientSideComponentId  799883f5-7962-4384-a10a-105adaec6ffc --clientSideComponentProperties '{ "someProperty": "Some value" }' --webUrl https://contoso.sharepoint.com/sites/sales
````
Let's remove double space in the value of the `clientSideComponentId` option

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.ts:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    try {
```
We shouldn't use a try..catch clause here, because we're using `Cli.executeCommand` without issuing additional web requests. `Cli.executeCommand` throws on failure `CommandError` exception which is handled by the CLI runtime without any additional work.

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,126 @@
+  const clientSideComponentId = '76d5f8c8-6228-4df8-a2da-b94cbc8115bc';
+  const clientSideComponentProperties = '';
+
+
```
Let's remove this unnecessary empty line

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,126 @@
+  before(() => {
+    sinon.stub(auth, 'restoreAuth').callsFake(() => Promise.resolve());
+    sinon.stub(telemetry, 'trackEvent').callsFake(() => { });
+    auth.service.connected = true;
```
We're missing here: `sinon.stub(pid, 'getProcessName').callsFake(() => '');`. We should include it to avoid unnecessary resolving of pid for the test runner

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,126 @@
+    assert.notStrictEqual(command.description, null);
+  });
+
+  it('adds application customizer to a specific site without specifying clientSideComponentId', async () => {
```
This test name is confusing, because it makes it sound like the command will add an app customizer without the ID, while in reality it won't. Also, this command would never run because it would fail validation. Looking at the implementation, are we testing here adding app customizer without properties rather than without the ID?

MathijsVerbeeck replied:
This should have been clientSideComponentProperties, my bad!

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,126 @@
+        return;
+      }
+
+      throw 'Invalid request';
```
This exception is odd since we're not executing a request but rather a command. Let's adjust it.

MathijsVerbeeck replied:
Adjusted the error message. In the 'throws an error' test, I will throw an actual error that I've retrieved when executing the command.

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,126 @@
+
+  it('adds application customizer to a specific site without specifying clientSideComponentId', async () => {
+    sinon.stub(Cli, 'executeCommand').callsFake(async (command) => {
+      if (command === SpoCustomActionAddCommand) {
```
If the only thing that we're doing in the command's code is executing a command, and here we circumvent it, what are we testing exactly?

MathijsVerbeeck replied:
I think that the command should make more sense now.

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,126 @@
+      throw 'Invalid request';
+    });
+
+    await command.action(logger, { options: { webUrl: webUrl, title: title, clientSideComponentId: clientSideComponentId, verbose: true } } as any);
```
Looking at the previous test case, the only difference here is `verbose: true`. In the `if` clause we're checking for clientSideComponentProperties but those will never be set because we're not passing them here, so that check won't do anything.

MathijsVerbeeck replied:
This is the same case, this should have been `clientSideComponentProperties`. There should be two test-cases:
1. Specifying `clientSideComponentProperties`
2. Not specifying `clientSideComponentProperties`

waldekmastykarz commented on src/m365/spo/commands/web/web-applicationcustomizer-add.ts:
```diff
@@ -0,0 +1,105 @@
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
```
Let's also check that the specified `clientSideComponentProperties` are valid JSON

waldekmastykarz reviewed (changes requested):
Great start! Let's do a few adjustments before we proceed.

martinlingstuyl commented:
Hi @MathijsVerbeeck, we've slightly updated the specs and removed the `web` from `spo web applicationcustomizer add`.
This is because a `scope` property will be included eventually, and it align with how we do it in the customaction commands.

Could you align your PR with the updated spec? 
(For the scope option we'll create an extra issue.)

MathijsVerbeeck commented:
Hi @waldekmastykarz 

My apologies for the 'sloppy' job. Could have done this PR way better!

I've resolved your comments I believe. Could you please have another look?

@martinlingstuyl  I've also renamed the name of the command to `spo applicationcustomizer add` and updated the references in the docs to the same name. Also moved the files accordingly.

waldekmastykarz commented:
No worries @MathijsVerbeeck, we're in this together. I'll have another look asap

waldekmastykarz reviewed (changes requested):
We're getting there! Let's do a small adjustments on the tests before we merge 😊

MathijsVerbeeck commented on src/m365/spo/commands/applicationcustomizer/applicationcustomizer-add.spec.ts:
```diff
@@ -0,0 +1,125 @@
+  it('adds application customizer to a specific site without specifying clientSideComponentProperties', async () => {
+    sinon.stub(Cli, 'executeCommand').callsFake(async (command, args) => {
+      if (command === SpoCustomActionAddCommand && args.options["clientSideComponentProperties"] === '') {
+        return;
```
Do you mean something like this? I change the stub from Cli, 'executeCommand' to request, 'post' and verify the url/data as done below?

```ts 
sinon.stub(request, 'post').callsFake(async (opts) => {
      if (opts.url === 'https://contoso.sharepoint.com/_api/Web/UserCustomActions' 
        && opts.data['Location'] === 'ClientSideExtension.ApplicationCustomizer'
        && opts.data['ClientSideComponentId'] === clientSideComponentId 
        && opts.data['ClientSideComponentProperties'] === clientSideComponentProperties 
        && opts.data['Name'] === title) {
        return;
      }

      throw customActionError;
});
```

waldekmastykarz replied:
Correct. That would decouple the test from the implementation making it more future-proof.

MathijsVerbeeck commented:
I've resolved your final comments. Could you have (hopefully a final) look?

waldekmastykarz commented:
Will do. Thank you for such a quick turnaround.

martinlingstuyl commented:
> I've resolved your final comments. Could you have (hopefully a final) look?

Thanks for sticking with us on this one @MathijsVerbeeck, it's taken a while but the end result will be awesome 👍💪

waldekmastykarz commented on src/m365/spo/commands/applicationcustomizer/applicationcustomizer-add.ts:
```diff
@@ -0,0 +1,109 @@
+      clientSideComponentId: args.options.clientSideComponentId,
+      clientSideComponentProperties: args.options.clientSideComponentProperties || '',
+      location: 'ClientSideExtension.ApplicationCustomizer'
+    };
```
We should pass verbose and debug options here, so that when user requests either output mode, we show that information. We should also pass output so that the command will return output in the specified output mode.

When I add these settings, I'm seeing the `adds application customizer to a specific site while specifying clientSideComponentProperties` test failing with `Cannot read properties of undefined (reading 'ClientSideComponentId')`. Could you please have a look at what's wrong?

MathijsVerbeeck replied:
Apparently, the command logs the output of the customaction add response when passing verbose. This is why. It has been solved now.

waldekmastykarz reviewed (changes requested):
We're almost there @MathijsVerbeeck! Let's have a final pass and we should be able to get it in 💪

waldekmastykarz reviewed (approved):
Nicely done! Nothing more to add 👏

waldekmastykarz commented:
Merged manually. Thank you! 👏
